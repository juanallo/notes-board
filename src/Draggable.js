import React from 'react';
import styled, { css } from 'styled-components';

export default class Draggable extends React.Component {

	constructor(props){
		super(props);

		this.state = {
			isDragging: false,
			x: props.x  || 0,
			y: props.y || 0,
			originalX: 0,
			originalY: 0,
		};
	}

	componentWillUnmount() {
		window.removeEventListener('mousemove', this.handleMouseMove);
		window.removeEventListener('mouseup', this.handleMouseUp);
	}

	handleMouseDown = ({ clientX, clientY }) => {
		window.addEventListener('mousemove', this.handleMouseMove);
		window.addEventListener('mouseup', this.handleMouseUp);

		if (this.props.onDragStart) {
			this.props.onDragStart();
		}

		this.setState({
			originalX: clientX,
			originalY: clientY,
			isDragging: true
		});
	};

	handleMouseMove = ({ clientX, clientY }) => {
		const { isDragging } = this.state;

		if (!isDragging) {
			return;
		}

		this.setState(prevState => {
			const mouseDx = clientX - prevState.originalX;
			const mouseDy = clientY - prevState.originalY;

			return {
				x: this.props.x + mouseDx,
				y: this.props.y + mouseDy
			}
		});
	};

	handleMouseUp = () => {
		window.removeEventListener('mousemove', this.handleMouseMove);
		window.removeEventListener('mouseup', this.handleMouseUp);

		this.setState(
			{
				originalX: 0,
				originalY: 0,
				isDragging: false
			},
			() => {
				if (this.props.onDragEnd) {
					this.props.onDragEnd({x: this.state.x, y: this.state.y});
				}
			}
		);
	};

	render() {
		const { children } = this.props;
		const { x, y, isDragging } = this.state;

		return (
			<Container
				onMouseDown={this.handleMouseDown}
				x={x}
				y={y}
				isDragging={isDragging}>
				{children}
			</Container>
		);
	}
}

const Container = styled.div.attrs({
	style: ({ x, y }) => ({
		transform: `translate(${x}px, ${y}px)`,
		position: 'relative'
	}),
})`
  cursor: grab;
  
  ${({ isDragging }) =>
	isDragging && css`
    opacity: 0.8;
    cursor: grabbing;
  `};
`;