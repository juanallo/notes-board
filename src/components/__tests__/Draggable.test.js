import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { mount, shallow, configure } from 'enzyme';
import Draggable from '../Draggable';

configure({adapter: new Adapter()});


it('load a Draggable in a particular position', () => {
	const wrapper = shallow(<Draggable x={0} y={10} />);
	expect(wrapper).toMatchSnapshot();
});

it('Mouse move while not dragging', () => {
	const wrapper = mount(<Draggable x={0} y={10} />);

	mouseMove(10, 20);

	expect(wrapper.state().x).toBe(0);
	expect(wrapper.state().y).toBe(10);
	expect(wrapper.state().isDragging).toBeFalsy();
});

it('Drag and drop a draggable', () => {
	const onDragEnd = jest.fn();
	const wrapper = mount(<Draggable x={0} y={10} onDragEnd={onDragEnd} />);
	wrapper.simulate("mousedown", {clientX: 0, clientY: 10});
	mouseMove(10, 20);

	expect(wrapper.state().x).toBe(10);
	expect(wrapper.state().y).toBe(20);

	mouseMove(30, 40);

	expect(wrapper.state().x).toBe(30);
	expect(wrapper.state().y).toBe(40);

	mouseUp();

	expect(onDragEnd.mock.calls.length).toBe(1);
	expect(onDragEnd.mock.calls[0][0]).toEqual({x: 30, y: 40});
});

it('Special l&f when dragging', () => {

});

function mouseMove(x, y) {
	const evt = document.createEvent('MouseEvents');
	evt.initMouseEvent("mousemove", true, true, window, 0, 0, 0, x, y, false, false, false, false, 0, null);
	document.dispatchEvent(evt);
	return evt;
}

function mouseUp(){
	const evt = document.createEvent('MouseEvents');
	evt.initMouseEvent("mouseup", true, true, window, 0, 0, 0, 0, 0, false, false, false, false, 0, null);
	document.dispatchEvent(evt);
}