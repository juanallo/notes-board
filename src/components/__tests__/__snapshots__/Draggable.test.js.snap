// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Draggable Test load a Draggable in a particular position 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Draggable
    x={0}
    y={10}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "checkPropTypes": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": undefined,
      "isDragging": false,
      "onMouseDown": [Function],
      "x": 0,
      "y": 10,
    },
    "ref": null,
    "rendered": null,
    "type": Object {
      "$$typeof": Symbol(react.forward_ref),
      "attrs": Array [
        Object {
          "style": [Function],
        },
      ],
      "componentStyle": ComponentStyle {
        "componentId": "sc-bdVaJa",
        "isStatic": false,
        "rules": Array [
          "
  cursor: grab;
  
  ",
          [Function],
          ";
",
        ],
      },
      "displayName": "styled.div",
      "foldedComponentIds": Array [],
      "render": [Function],
      "styledComponentId": "sc-bdVaJa",
      "target": "div",
      "toString": [Function],
      "warnTooManyClasses": [Function],
      "withComponent": [Function],
    },
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": undefined,
        "isDragging": false,
        "onMouseDown": [Function],
        "x": 0,
        "y": 10,
      },
      "ref": null,
      "rendered": null,
      "type": Object {
        "$$typeof": Symbol(react.forward_ref),
        "attrs": Array [
          Object {
            "style": [Function],
          },
        ],
        "componentStyle": ComponentStyle {
          "componentId": "sc-bdVaJa",
          "isStatic": false,
          "rules": Array [
            "
  cursor: grab;
  
  ",
            [Function],
            ";
",
          ],
        },
        "displayName": "styled.div",
        "foldedComponentIds": Array [],
        "render": [Function],
        "styledComponentId": "sc-bdVaJa",
        "target": "div",
        "toString": [Function],
        "warnTooManyClasses": [Function],
        "withComponent": [Function],
      },
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "legacyContextMode": "parent",
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getChildContext": Object {
            "calledByRenderer": false,
          },
          "getDerivedStateFromProps": Object {
            "hasShouldComponentUpdateBug": false,
          },
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
  Symbol(enzyme.__childContext__): null,
}
`;
